### PR TITLE
Node.js basics

### PR DESCRIPTION
**Pull request by Arterixs**
1. Task: https://github.com/AlreadyBored/nodejs-assignments/blob/main/assignments/nodejs-basics/assignment.md
2. Done 17.06.2023 / deadline 20.06.2023
3. Score: 206/206
 - File system (src/fs)
  -[x] ```create.js``` implemented properly (6)
  -[x] ```copy.js``` implemented properly (10)
  -[x] ```rename.js``` implemented properly (10)
  -[x] ```delete.js``` implemented properly (6)
  -[x] ```list.js``` implemented properly (6)
  -[x] ```read.js``` implemented properly (6)
- Command line interface(src/cli)
  -[x] ```env.js``` implemented properly (6)
  -[x] ```args.js``` implemented properly (6)
- Modules(src/modules)
  -[x] ```cjsToEsm.cjs``` refactored properly (20)
- Hash (src/hash)
  -[x] ```calcHash.js``` implemented properly (10)
- Streams (src/streams)
  -[x] ```read.js``` implemented properly (10)
  -[x] ```write.js``` implemented properly (10)
  -[x] ```transform.js``` implemented properly (10)
- Zlib (src/zip)
  -[x] ```compress.js``` implemented properly (10)
  -[x] ```decompress.js``` implemented properly (10)
- Worker Threads (src/wt)
  -[x] ```worker.js``` implemented properly (10)
  -[x] ```main.js``` implemented properly (30)
- Child Processes (src/cp)
  -[x] spawns child process (10)
  -[x] child process stdin receives input from master process stdin (10)
  -[x] child process stdout sends data to master process stdout (10)

### CODE REVIEW

File: src/helpers/index.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { fileURLToPath } from "node:url";
+import path, { dirname } from "node:path";
+
+export const getPathCurrent = (url, ...args) => {
+  const _filename = fileURLToPath(url);
+  return path.join(dirname(_filename), ...args);
+};
+
+export const triggerErrorFs = (error) => {
```
Comment (morphlne) [inline src/helpers/index.js:9]: `triggerErrorFs` does not reflect essence. `rethrowAsFsError`?

File: src/fs/create.js [hunk 2]
```diff
@@ -1,5 +1,15 @@
+import { writeFile } from "node:fs/promises";
+import { getPathCurrent, triggerErrorFs } from "../helpers/index.js";
+
 const create = async () => {
-    // Write your code here 
+  const path = getPathCurrent(import.meta.url, "files", "fresh.txt");
+  try {
```
Comment (morphlne) [inline src/fs/create.js:6]: Promise treats every error as rejection (even thrown inside `.catch`), so `try... catch` won't work as expected and catch block will never be reached

Adding `await` before `writeFile` does the trick

File: src/fs/copy.js [hunk 3]
```diff
@@ -1,5 +1,27 @@
...
+  const pathOriginMcdir = getPathCurrent(import.meta.url, "files");
+  try {
+    const files = await readdir(pathOriginMcdir).catch((error) =>
+      triggerErrorFs(error)
+    );
+    await mkdir(pathCopyMcdir, { recursive: false }).catch((error) =>
+      triggerErrorFs(error)
+    );
+    files.forEach(async (item) => {
+      if (item.includes(".")) {
```
Comment (morphlne) [inline src/fs/copy.js]: Why filter source content?

File: src/fs/copy.js [hunk 4]
```diff
@@ -1,5 +1,27 @@
...
+        await copyFile(
```
Comment (morphlne) [inline src/fs/copy.js:16]: `await` inside `forEach` is definitely not the good practice

File: src/helpers/index.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { fileURLToPath } from "node:url";
+import path, { dirname } from "node:path";
+
+export const getPathCurrent = (url, ...args) => {
```
Comment (morphlne) [inline src/helpers/index.js:4]: Every time `import.meta.url` is passed as `url`. Maybe encapsulate `url` in helper file?

File: src/fs/copy.js [hunk 6]
```diff
@@ -1,5 +1,27 @@
...
+
 const copy = async () => {
-    // Write your code here 
+  const pathCopyMcdir = getPathCurrent(import.meta.url, "files_copy");
+  const pathOriginMcdir = getPathCurrent(import.meta.url, "files");
+  try {
+    const files = await readdir(pathOriginMcdir).catch((error) =>
+      triggerErrorFs(error)
+    );
+    await mkdir(pathCopyMcdir, { recursive: false }).catch((error) =>
```
Comment (morphlne) [inline src/fs/copy.js:11]: Can be run in parallel with `readdir`?

File: src/fs/rename.js [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+    "wrongFilename.txt"
+  );
+  const pathNewFile = getPathCurrent(
+    import.meta.url,
+    "files",
+    "properFilename.md"
+  );
+  try {
+    renamed(pathOldFile, pathNewFile).catch((err) => triggerErrorFs(err));
+  } catch (err) {
```
Comment (morphlne) [inline src/fs/rename.js:21]: `catch` will never be reached

File: src/fs/rename.js [hunk 8]
```diff
@@ -1,5 +1,22 @@
...
+    "files",
+    "wrongFilename.txt"
+  );
+  const pathNewFile = getPathCurrent(
+    import.meta.url,
+    "files",
+    "properFilename.md"
+  );
+  try {
+    renamed(pathOldFile, pathNewFile).catch((err) => triggerErrorFs(err));
```
Comment (morphlne) [inline src/fs/rename.js]: No error thrown if both files exist

File: src/fs/create.js [hunk 9]
```diff
@@ -1,5 +1,15 @@
+import { writeFile } from "node:fs/promises";
+import { getPathCurrent, triggerErrorFs } from "../helpers/index.js";
```
Comment (morphlne) [inline src/fs/create.js]: `index` is not obvious name for utility file

File: src/fs/list.js [hunk 10]
```diff
@@ -1,5 +1,15 @@
...
+import { getPathCurrent, triggerErrorFs } from "../helpers/index.js";
+
 const list = async () => {
-    // Write your code here 
+  const pathMcdir = getPathCurrent(import.meta.url, "files");
+  try {
+    readdir(pathMcdir)
+      .then((result) => console.log(result))
+      .catch((error) => triggerErrorFs(error));
+  } catch (err) {
```
Comment (morphlne) [inline src/fs/list.js:11]: The same...

File: src/fs/list.js [hunk 11]
```diff
@@ -1,5 +1,15 @@
+import { readdir } from "node:fs/promises";
+import { getPathCurrent, triggerErrorFs } from "../helpers/index.js";
+
 const list = async () => {
-    // Write your code here 
+  const pathMcdir = getPathCurrent(import.meta.url, "files");
+  try {
+    readdir(pathMcdir)
+      .then((result) => console.log(result))
```
Comment (morphlne) [inline src/fs/list.js]: `result` contains dir names too

File: src/fs/read.js [hunk 12]
```diff
@@ -1,5 +1,15 @@
+import { readFile } from "node:fs/promises";
+import { getPathCurrent, triggerErrorFs } from "../helpers/index.js";
+
 const read = async () => {
-    // Write your code here 
+  const pathMcdir = getPathCurrent(import.meta.url, "files", "fileToRead.txt");
```
Comment (morphlne) [inline src/fs/read.js]: `McDir` ?

File: src/streams/read.js [hunk 13]
```diff
@@ -1,5 +1,17 @@
...
 const read = async () => {
-    // Write your code here 
+  const pathReadFile = getPathCurrent(
+    import.meta.url,
+    "files",
+    "fileToRead.txt"
+  );
+  const readFile = createReadStream(pathReadFile);
+  readFile.on("data", function (chunk) {
+    stdout.write(chunk.toString());
```
Comment (morphlne) [inline src/streams/read.js]: `toString` can be omitted

File: src/streams/transform.js [hunk 14]
```diff
@@ -1,5 +1,17 @@
+import { Transform } from "node:stream";
+import { pipeline } from "node:stream/promises";
+import { stdin, stdout } from "node:process";
+
 const transform = async () => {
-    // Write your code here 
+  const reverseTransform = new Transform({
+    transform(chunk, encoding, callback) {
+      const cleanString = chunk.toString().trim();
```
Comment (morphlne) [inline src/streams/transform.js]: `.trim()` also removes spaces `abc   ` => `cba`, but `   cba` is expected

File: src/streams/transform.js [hunk 15]
```diff
@@ -1,5 +1,17 @@
...
+      const reverseString = cleanString.split("").reverse().join("");
+      const finalResult = `${reverseString}\n`;
```
Comment (morphlne) [inline src/streams/transform.js]: EOL is preferred over os-dependent
Please take a look https://nodejs.org/dist/latest-v18.x/docs/api/os.html#oseol

File: src/streams/transform.js [hunk 16]
```diff
@@ -1,5 +1,17 @@
...
+      callback(null, finalResult);
+    },
+  });
+  await pipeline(stdin, reverseTransform, stdout);
```
Comment (morphlne) [inline src/streams/transform.js:15]: Can be `await` omitted?

File: src/wt/worker.js [hunk 17]
```diff
@@ -1,8 +1,11 @@
...
+  n < 2 ? n : nthFibonacci(n - 1) + nthFibonacci(n - 2);
+
+const sendResult = (number) => {
+  const result = nthFibonacci(number);
+  parentPort.postMessage(result);
 };
 
-sendResult();
\ No newline at end of file
+sendResult(workerData);
```
Comment (morphlne) [inline src/wt/worker.js:11]: From requirements:

- Don't change signature of pre-written functions (e.g. don't rename them, don't make them synchronous, etc.)

File: src/wt/main.js [hunk 18]
```diff
@@ -1,5 +1,36 @@
...
+      worker.on("message", (value) => resolve(value));
+      worker.on("messageerror", () => reject());
+      worker.on("error", () => reject());
+    });
+    arrPromisesWorker.push(promiseWorker);
+  }
+
+  Promise.allSettled(arrPromisesWorker)
+    .then((result) => {
+      const updateResult = result.map((item) => {
```
Comment (morphlne) [inline src/wt/main.js:25]: `return`?

File: src/fs/files/fileToRemove.txt [hunk 19]
```diff
@@ -1 +0,0 @@
-How dare you!
```
Comment (morphlne) [inline src/fs/files/fileToRemove.txt:1]: 🤔

File: src/modules/files/a.json [hunk 20]
```diff
@@ -1,5 +1,5 @@
 {
-    "a": 1,
```
Comment (morphlne) [inline src/modules/files/a.json:2]: 🤔

**Review by morphlne — COMMENTED** [review]
Hi Arterixs

All tasks work well, alongside I think some moments can be improved.
Regarding score i provided detailed overview in RSS app.

Excellent job!

File: src/helpers/index.js (hunk 1 above)
Comment (Arterixs) [reply src/helpers/index.js:9]: Ok, thanks

**Pull request merged**